### PR TITLE
Fixed Android - Chat - Message gets displayed from right to left

### PR DESCRIPTION
**Pull request by samilabud**
### Details

The root cause is that Android only converts RTL text to LTR text using Unicode controls and by default, when you start writing in a language that is written in RTL mode it converts the input text to RTL.

### Fixed Issues
$ https://github.com/Expensify/App/issues/30584
PROPOSAL:  https://github.com/Expensify/App/issues/30584#issuecomment-1785494092

### Tests
1. Launch App
2. Open any chat and paste an RTL text in the compose box, for example " مثال "
3. Continue typing some content

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
1. Launch App
2. Open any chat and paste an RTL text in the compose box, for example " مثال "
3. Continue typing some content

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/5216036/8dbd60b9-d3c2-4887-a8bf-2539811c25b0

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/5216036/88df71ff-8445-4ab3-b3ce-aedeb4013516

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/5216036/2d00cef0-171a-471a-9afc-d7bed970975e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/5216036/0227ccf4-78bc-46f8-a830-9ef00d60ddac

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/5216036/4ea4afd5-cc9d-42c4-bd97-113f3a476b58

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/5216036/fb19538c-46b6-409f-bd3d-cb938afd42cc

</details>

### CODE REVIEW

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 1]
```diff
@@ -523,6 +524,12 @@ function ComposerWithSuggestions({
         }),
         [blur, focus, prepareCommentAndResetComposer, replaceSelectionWithText],
     );
+    // eslint-disable-next-line rulesdir/prefer-early-return
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js]: Done!

**Comment by HardikChoudhary24** [comment]
@samilabud @allroundexperts @tgolen  As I had worked on this issue for other platforms previously, I was getting some issues, and similar issues are visible in this change also:

- Mention Suggestions
- Draft comments
- Sending an empty message


https://github.com/Expensify/App/assets/126149345/3c297111-24df-4f36-b490-aadcc259c150

**Comment by allroundexperts** [comment]
Thanks for testing this out @HardikChoudhary24. In the screenshots I attached, I can see that the placeholder was appearing at-least 🤔. @samilabud Can you please check the issues that @HardikChoudhary24 pointed out?

**Comment by samilabud** [comment]
Good catch @HardikChoudhary24 thank you!. 

@allroundexperts currently this works better, please check it out:

https://github.com/Expensify/App/assets/5216036/9c8ee927-cc4a-4eb5-87c9-9161a6c90dc7

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 2]
```diff
@@ -105,6 +106,7 @@ function ComposerWithSuggestions({
     const styles = useThemeStyles();
     const {preferredLocale} = useLocalize();
     const isFocused = useIsFocused();
+    const [firstChange, setFirstChange] = useState(true);
```
Comment (tgolen) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:109]: Is there a particular reason you are using `useState` instead of `useRef`? Since `firstChange` isn't used in rendering, I think using a ref would be more appropriate.

Comment (tgolen) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:109]: `firstChange` isn't a very good variable name. What about something like `hasCommentBeenUpdatedOnce`? Though, I'm not really following the code about why this variable is necessary or what the intention of it is.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 3]
```diff
@@ -220,16 +222,42 @@ function ComposerWithSuggestions({
...
+            let newCommentConverted = convertToLTRForComposer(newComment);
+
+            const moveCursorToEndOfLine = (commentLength) => {
+                setSelection({
+                    start: commentLength + 1,
+                    end: commentLength + 1,
+                });
+            };
+
+            if (firstChange && commentRef.current === '' && !['@'].includes(newComment)) {
```
Comment (tgolen) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:235]: Please add code comments to explain _why_ all this logic is here.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 4]
```diff
@@ -220,16 +222,42 @@ function ComposerWithSuggestions({
...
+            }
+
             const isNewCommentEmpty = !!newCommentConverted.match(/^(\s)*$/);
             const isPrevCommentEmpty = !!commentRef.current.match(/^(\s)*$/);
 
             /** Only update isCommentEmpty state if it's different from previous one */
             if (isNewCommentEmpty !== isPrevCommentEmpty) {
                 setIsCommentEmpty(isNewCommentEmpty);
+                if (isNewCommentEmpty) {
+                    setFirstChange(true);
```
Comment (tgolen) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:255]: It's strange to see this being set here, but maybe it's only confusing because of the variable name.

**Review by tgolen — CHANGES_REQUESTED** [review]
I feel like the diff changed by an order of magnitude since the last time I looked at it and now I don't really follow the scope of the solution.

Are the problems mentioned by @HardikChoudhary24 caused by your PR, or are they existing bugs that are being fixed in this PR?

I think if they are regressions caused by your PR, they should be fixed in this PR. If they are existing bugs, then we should open up a new GH to track and fix them independently.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 3 above)
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:235]: Got it, done!

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 2 above)
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:109]: Good point, I think `composerIsEmpty` could make more sense, please let me know what you think.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 4 above)
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:255]: My bad, we need to specify when the composer is empty because when unicode is present It is not really empty but it detects like it does.

**Comment by samilabud** [comment]
> I feel like the diff changed by an order of magnitude since the last time I looked at it and now I don't really follow the scope of the solution.
> 
> Are the problems mentioned by @HardikChoudhary24 caused by your PR, or are they existing bugs that are being fixed in this PR?
> 
> I think if they are regressions caused by your PR, they should be fixed in this PR. If they are existing bugs, then we should open up a new GH to track and fix them independently.

Hi, issues mentioned by @HardikChoudhary24 are caused by my PR but it looks like they tried this before and didn't works or something.

I have fixed every scenario inside the corresponding files to make this change affects only android mobiles, please check it out 🙏🏼

File: src/libs/convertToLTR/index.android.ts [hunk 5]
```diff
@@ -7,4 +7,30 @@ import ConvertToLTR from './types';
...
+    ) => void,
+) => {
+    setSelection({
+        start: commentLength + 1,
+        end: commentLength + 1,
+    });
+};
+
+/** We should remove the LTR unicode when the input is empty to prevent: Sending an empty message, bad function of metion suggestions (force option: always remove the unicode), or placeholder issues */
+const removeUnicodeLTRWhenEmpty = (newComment: string, newCommentConverted: string, force?: boolean) => {
```
Comment (tgolen) [inline src/libs/convertToLTR/index.android.ts:27]: There are a few things I am noticing for this method.

1. It is not clear when `force` should be used
2. I don't know what "bad function of metion suggestions" means or implies (it also has a typo)
3. The name `removeUnicodeLTRWhenEmpty` implies that the the world outside of this lib knows implementation details from inside the lib, primarily the unicode character. I think this should be renamed to something more like `resetLTRWhenEmpty`
4. The name `newComment` and `newCommentConverted` implies that this lib has knowledge of the outside world and how it is used, which is bad for a library. Can this be more generic and be called `inputString` and `inputStringConverted`?
5. Why is the `newCommentConverted` variable necessary? It's strange to see and isn't intuitive.

File: src/libs/convertToLTR/index.ts [hunk 6]
```diff
@@ -2,4 +2,18 @@ import ConvertToLTR from './types';
 
 const convertToLTR: ConvertToLTR = (text) => text;
 
+const moveCursorToEndOfLine = (
```
Comment (tgolen) [inline src/libs/convertToLTR/index.ts:5]: Can you please add a comment to the top of this file that explains the difference between Android and the rest of the platforms? Something like this should be fine:

> The Android platform has to handle switching between LTR and RTL languages a bit differently (see the Android file for more details). For all other platforms, these can simply be no-op functions.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 7]
```diff
@@ -220,26 +222,45 @@ function ComposerWithSuggestions({
...
             }
-            const newCommentConverted = convertToLTRForComposer(newComment);
+
+            let newCommentConverted = newComment;
+            const prevComment = commentRef.current;
+
+            // This prevent the double execution of setting input value that could affect the place holder and could send an empty message or draft messages in android
+            if (prevComment !== newComment) {
+                newCommentConverted = removeUnicodeLTRWhenEmpty(newComment, newCommentConverted);
+                newCommentConverted = convertToLTRForComposer(newCommentConverted, composerIsEmpty);
```
Comment (tgolen) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:232]: There are two methods being called in succession, with the variable being overwritten each time. It seems like these should be combined somehow (maybe using a parameter to clear the unicode character when empty).

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 8]
```diff
@@ -220,26 +222,45 @@ function ComposerWithSuggestions({
                     debouncedUpdateFrequentlyUsedEmojis();
                 }
             }
-            const newCommentConverted = convertToLTRForComposer(newComment);
+
+            let newCommentConverted = newComment;
```
Comment (tgolen) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:226]: I suggest renaming this variable to `newCommentConvertedToLTR` since "converted" is ambiguous.

**Review by tgolen — CHANGES_REQUESTED** [review]
I know this is a lot of comments, but the changes in this PR really went down an interesting path!

File: src/libs/convertToLTR/index.android.ts [hunk 9]
```diff
@@ -7,4 +7,26 @@ import ConvertToLTR from './types';
...
+    end: number;
+}>)=>void) => {
+    setSelection({
+        start: commentLength + 1,
+        end: commentLength + 1,
+    });
+}
+
+/** We should remove the LTR unicode when the input is empty to prevent: Sending an empty message, bad function of metion suggestions (force option: always remove the unicode), or placeholder issues */
+const removeUnicodeLTRWhenEmpty = (newComment:string, newCommentConverted:string, force?:boolean) => {
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/libs/convertToLTR/index.android.ts]: Oh, I see some of my comments posted twice. Sorry about that. Some of the comments disappeared when there was an update to the branch.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 10]
```diff
@@ -220,26 +222,45 @@ function ComposerWithSuggestions({
...
+            const prevComment = commentRef.current;
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:227]: Here I just put that in a variable, it could work but I prefer using the same pattern that was used (just to let everything be like before), e.g:

https://github.com/Expensify/App/blob/2feaac80198481546ee6aa630aa1c7b386ec5605/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L225


https://github.com/Expensify/App/blob/2feaac80198481546ee6aa630aa1c7b386ec5605/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L234

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 8 above)
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:226]: I can change it, but it is the original name:

https://github.com/Expensify/App/blob/2feaac80198481546ee6aa630aa1c7b386ec5605/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L223

Please confirm 🙏🏼

File: src/libs/convertToLTR/index.ts (hunk 6 above)
Comment (samilabud) [reply src/libs/convertToLTR/index.ts:5]: I haven't found any official documentation of this, except for this one https://developer.android.com/training/basics/supporting-devices/languages, I have added the comment.

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 11]
```diff
@@ -1,8 +1,9 @@
+import CONST from '@src/CONST';
 import ConvertToLTRForComposer from './types';
 
 /**
  * Android only - Do not convert RTL text to a LTR text for input box using Unicode controls.
  * Android does not properly support bidirectional text for mixed content for input box
  */
-const convertToLTRForComposer: ConvertToLTRForComposer = (text) => text;
+const convertToLTRForComposer: ConvertToLTRForComposer = (text, isComposerEmpty) => (isComposerEmpty ? `${CONST.UNICODE.LTR}${text}` : text);
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/libs/convertToLTRForComposer/index.android.ts:8]: I am not sure if I understand the question but, it is one for Android and the other for the rest of the devices.

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js [hunk 12]
```diff
@@ -105,6 +106,7 @@ function ComposerWithSuggestions({
     const styles = useThemeStyles();
     const {preferredLocale} = useLocalize();
     const isFocused = useIsFocused();
+    const [composerIsEmpty, setComposerIsEmpty] = useState(true);
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:109]: Changed to using ref, thank you!

File: src/libs/convertToLTR/index.android.ts [hunk 13]
```diff
@@ -7,4 +7,26 @@ import ConvertToLTR from './types';
...
+    
+    const removeRegEx = new RegExp(`${CONST.UNICODE.LTR}`,'g')
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/libs/convertToLTR/index.android.ts]: Good point, removed, thanks

**Comment by samilabud** [comment]
> I know this is a lot of comments, but the changes in this PR really went down an interesting path!

Thanks for the suggestions, I think I've addressed everything, please check back when you get a chance 🙏🏼

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 8 above)
Comment (tgolen) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:226]: Yes, please change it. Thanks for confirming.

File: src/libs/convertToLTRForComposer/index.android.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (tgolen) [reply src/libs/convertToLTRForComposer/index.android.ts]: What I meant was that we have a lib called `convertToLTR` and another one `convertToLTRForComposer`. Can those be consolidated into a single lib?

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js (hunk 8 above)
Comment (samilabud) [reply src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js:226]: Done!

**Comment by samilabud** [comment]
Hi @tgolen, I think I finished with the changes please double-check because I marked the conversations as resolved and currently I see just 2 comments from you, could you please mark them as unsolved when you add a new comment? or maybe it is better to never mark comments as resolved 🤔.

BTW when I was testing I found a new bug (this is not related to this PR, the test was in https://staging.new.expensify.com):

https://github.com/Expensify/App/assets/5216036/e5ea72d0-d6ac-4216-b5d6-4cae07f89f14

Do you know where we can report this? I think we can fix it too.

**Comment by tgolen** [comment]
For bugs not related to your PR, you can post them in the slack channel #expensify-bugs. 

For the comments, I would usually never manually mark things as resolved. I know there were some issues with my previous review and having multiple comments, so I will go through it this morning and clean it up and give another review.

**Review by tgolen — DISMISSED** [review]

**Comment by tgolen** [comment]
@allroundexperts do you want to give another testing pass on this to ensure it's still in working order before it's merged? There were quite a few changes since your initial testing I think.

**Comment by allroundexperts** [comment]
Sure thing @tgolen. I'll retest this tonight/tomorrow.

**Comment by samilabud** [comment]
> Looks much better @samilabud. However, I still got the following bug:
> 
>  Screen.Recording.2023-12-11.at.6.34.15.AM.mov

Thank you very much for testing this out, it was a great catch! it is fixed now:

https://github.com/Expensify/App/assets/5216036/90bf4e0d-679f-426b-b934-aa55d423dbf6

**Review by tgolen — DISMISSED** [review]

**Comment by tgolen** [comment]
It looks like this is all approved and everything has been addressed, but I don't want to merge it too early if I'm missing something. @allroundexperts Is this OK to merge now?

**Comment by allroundexperts** [comment]
Yes, last time I reviewed this, I found a bug. I need to test it again.

**Comment by allroundexperts** [comment]
@samilabud Still getting the following error:

https://github.com/Expensify/App/assets/30054992/d8a85afa-bfb9-44bd-887f-fdcdc85e8ccf

At this point, since we have deviated so much from the original proposal, I'm wondering if its best to close this PR and go back to the drawing board again?

**Comment by samilabud** [comment]
> @samilabud Still getting the following error:
> 
>  Screen.Recording.2023-12-13.at.12.50.41.AM.mov 
> At this point, since we have deviated so much from the original proposal, I'm wondering if its best to close this PR and go back to the drawing board again?

Hey, what is the bug? I can't see the issue, maybe this is the wrong video 😅, please confirm 🙏🏼

**Comment by allroundexperts** [comment]
@samilabud Updated with correct video 😄

**Comment by samilabud** [comment]
Hi team, I think is fixed now, the issue was because we should determine when we should convert the composer to LTR, I have used a regular expression for each of the cases we already know, and this is the result:

https://github.com/Expensify/App/assets/5216036/f4e3ac4a-b4b7-41ce-bb05-cfd40b53473c

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 14]
```diff
@@ -1,8 +1,53 @@
+import CONST from '@src/CONST';
 import ConvertToLTRForComposer from './types';
 
+/**
+ * Android only - We need to determinate when composer is considered empty this going to be used when we don't want to convert the input composer to LTR
+ */
+function hasBeenComposerConsideredEmpty(text: string): boolean {
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:7]: I think a better name would be `canComposerBeConvertedToLTR`

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 15]
```diff
@@ -1,8 +1,53 @@
...
+    // Regular expressions with all character that are consider as empty (spaces, unicode with spaces and unicode with at)
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:8]: Please update the comment to explain the empty expressions better. It might even be better to use three different variables so that it's more clear what the regexes are for. I'm looking for comments like this:

```
// The composer is considered "empty" if all it contains is the LTR character followed by an @. This handles
// the case where someone has RTL enabled and they began typing an @mention for someone. Without this,
// it leads to behavior where ... blah blah
const ltrRegex = new RegExp(`^${CONST.UNICODE.LTR}@$`);
```

With comments of this nature, it would be nearly impossible for someone to misinterpret why that specific regex is being used.

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 16]
```diff
@@ -1,8 +1,53 @@
...
+    const emptyExpressions = [/^\s*$/, new RegExp(`^${CONST.UNICODE.LTR}@$`), new RegExp(`${CONST.UNICODE.LTR}\\s*$`)];
+    return emptyExpressions.some((exp) => exp.test(text));
+}
+
+/**
+ * Android only - We should remove the LTR unicode when the input is empty to prevent: Sending an empty message, metion suggestions not works if @ or \s (at or space) is the first character; (force option: always remove the unicode, going to be used when composer is consider as empty) or placeholder not shows if unicode character is the only remaining character. */
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:14]: Shorten this comment so it is less than 190 characters wide, please.

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/libs/convertToLTRForComposer/index.android.ts (hunk 15 above)
Comment (samilabud) [reply src/libs/convertToLTRForComposer/index.android.ts:8]: Great suggestion thanks!, I have created the different variables and I have added comments to explain each one, please confirm.

File: src/libs/convertToLTRForComposer/index.android.ts (hunk 16 above)
Comment (samilabud) [reply src/libs/convertToLTRForComposer/index.android.ts:14]: My bad, fixed!

**Comment by samilabud** [comment]
Hi @tgolen, the suggestions were implemented, please check again.

**Comment by samilabud** [comment]
I did some tests after the last change and everything seems to be ok:

https://github.com/Expensify/App/assets/5216036/cf12206d-9ef8-4e34-8734-bf57a5a905b0

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 17]
```diff
@@ -1,8 +1,62 @@
...
+    const emptyExpressions = [containOnlySpaces, startsWithLTRAndAt, startsWithLTRAndSpace];
+    return emptyExpressions.some((exp) => exp.test(text));
+}
+
+/**
+ * Android only - We should remove the LTR unicode when the input is empty to prevent:
+ *  Sending an empty message;
+ *  Mention suggestions not works if @ or \s (at or space) is the first character;
+ *  Placeholder is not displayed if the unicode character is the only character remaining;
+ * force: always remove the LTR unicode, going to be used when composer is consider as empty */
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:23]: The `*/` needs to go on the next line by itself.

Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:23]: If the `force:` comment is for the force param, please format it as a legit `@param` doc.

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 18]
```diff
@@ -1,8 +1,62 @@
+import CONST from '@src/CONST';
 import ConvertToLTRForComposer from './types';
 
+/**
+ * Android only - The composer can be converted to LTR if its content is the LTR character followed by an @ or space
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:5]: This comment should also explain _why_ it's OK to convert it in these cases and _why_ it shouldn't be converted. Consider adding "because..." and then fill out the rest.

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 19]
```diff
@@ -1,8 +1,62 @@
...
+ */
+function canComposerBeConvertedToLTR(text: string): boolean {
+    // this handle cases when user type only spaces
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:8]: Please start comments with a capital letter

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 20]
```diff
@@ -1,8 +1,62 @@
...
+    const containOnlySpaces = /^\s*$/;
+    // this handle the case where someone has RTL enabled and they began typing an @mention for someone.
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:10]: Do comments end with a period or not? Let's be consistent

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 21]
```diff
@@ -1,8 +1,62 @@
...
+    const startsWithLTRAndAt = new RegExp(`^${CONST.UNICODE.LTR}@$`);
+    // this handle cases could send empty messages when composer is multiline
```
Comment (tgolen) [inline src/libs/convertToLTRForComposer/index.android.ts:12]: I'm not quite sure what this means and the grammar is a bit confusing. Are you trying to say this?

> This regex handles the case where the composer can contain multiple lines of whitespace.

File: src/libs/convertToLTRForComposer/index.android.ts (hunk 19 above)
Comment (tgolen) [reply src/libs/convertToLTRForComposer/index.android.ts:8]: "this handle cases" is not good grammar and should be more like "This regex handles the case when a user only types spaces into the composer.

**Review by tgolen — COMMENTED** [review]
It's looking better! I think we just need to clean up some of the grammar and comment formatting and then this is good.

File: src/libs/convertToLTRForComposer/index.android.ts (hunk 21 above)
Comment (samilabud) [reply src/libs/convertToLTRForComposer/index.android.ts]: That's right, sorry 😣!

**Comment by samilabud** [comment]
Hi @tgolen, please verify again when you get a chance.

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/libs/convertToLTRForComposer/index.android.ts [hunk 22]
```diff
@@ -1,8 +1,66 @@
...
+}
+
+/**
+ * Android only - We should remove the LTR unicode when the input is empty to prevent:
+ *  Sending an empty message;
+ *  Mention suggestions not works if @ or \s (at or space) is the first character;
+ *  Placeholder is not displayed if the unicode character is the only character remaining;
+ *
+ * @newComment: the comment written by the user
+ * @force: always remove the LTR unicode, going to be used when composer is consider as empty */
```
[the remark this answers is not in the document]
Comment (samilabud) [reply src/libs/convertToLTRForComposer/index.android.ts]: Sorry, thanks for the link, I appreciate it (TIL).
Done.

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
All yours @allroundexperts

**Comment by allroundexperts** [comment]
@samilabud Now I am getting the following issue:

https://github.com/Expensify/App/assets/30054992/a14fafd9-ac50-419e-9348-5d64527c7741

**Comment by samilabud** [comment]
> @samilabud Now I am getting the following issue:
> 
>  Screen.Recording.2023-12-19.at.7.13.21.PM.mov

Hi, do you mean the delay where text takes to be from left to right? If so, this seems to be the expect behavior due android has not transitions when we convert the input to LTR from RTL

**Comment by allroundexperts** [comment]
@samilabud I'm talking about the text suddenly moving from right to left. This did not happen in your previous versions. You can also confirm this by watching your / mine screen recordings for Android.

**Comment by samilabud** [comment]
> @samilabud I'm talking about the text suddenly moving from right to left. This did not happen in your previous versions. You can also confirm this by watching your / mine screen recordings for Android.

Got it, I have noticed this behavior from beginning but it seems related to emulator/device available resources, this happen even in web version with a bug I reported previously in this thread.

**Comment by samilabud** [comment]
@allroundexperts I'm thinking maybe you mean the way it was it the beginning that we never allow the input to be converted to RTL but this causes mentions to not work, or blank messages to be sent, and so on. If this is a big issue please let me know so I can try to find another way to convert the input.

**Comment by tgolen** [comment]
What's the status with this?

@allroundexperts I think you are suggesting that earlier commits in this PR didn't have the delay. Is that right?

**Comment by samilabud** [comment]
Yes, this has been happening since the first commits (see the video at sec 3 to [this comment](https://github.com/Expensify/App/pull/32297#issuecomment-1841112466)) but maybe you don't notice because this depends on the resources available:

image

The reason is that we previously converted (in the proposal) the input before the user started typing (adding the Unicode), and now we have to wait for the user to type (to add the Unicode if has the conditions - to avoid sending an empty message, or mentions and placeholder works correctly). 

In summary, as we have to wait for the user to write it is noticeable when Android converts the input from RTL to LTR.

**Comment by tgolen** [comment]
I think that is OK to live with for now. Thank you for explaining! @allroundexperts are you OK for me to merge this then?

**Comment by allroundexperts** [comment]
@tgolen If that's fine, then I think we can merge this.

**Comment by tgolen** [comment]
Late Christmas present :D I wish I could have merged this yesterday 🎄 🎅

**Comment by samilabud** [comment]
> Late Christmas present :D I wish I could have merged this yesterday 🎄 🎅

Thank you all, merry late Christmas! 🎄🎁 😊
